compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2037,912 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
+      generateCompareBranchSequence(trgReg, node, firstChild, secondChild, compareInfo, cg);
       }
+
    node->setRegister(trgReg);
    cg->decReferenceCount(firstChild);
    cg->decReferenceCount(secondChild);
    return trgReg;
    }
 
-TR::Register *OMR::Power::TreeEvaluator::lcmpleEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+TR::Register *OMR::Power::TreeEvaluator::fcmpeqEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   if (cg->comp()->target().is32Bit())
-      return compareLongAndSetOrderedBoolean(TR::InstOpCode::cmp4, TR::InstOpCode::ble, node, cg);
+   return floatCompareEvaluator(node, CompareInfo(CompareCondition::eq, TR::Float, false), cg);
+   }
 
-   TR::Node     *firstChild   = node->getFirstChild();
-   TR::Node     *secondChild  = node->getSecondChild();
-   TR::Register *src1Reg      = cg->evaluate(firstChild);
-   TR::Register *trgReg    = cg->allocateRegister();
-   if (secondChild->getOpCode().isLoadConst() &&
-       secondChild->getRegister() == NULL)
-      {
-      TR::Register    *temp1Reg;
-      int64_t value = secondChild->getLongInt();
-      bool    killTemp1 = false;
-      if (value == -1)
-         {
-         temp1Reg = src1Reg;
-         }
-      else
-         {
-         TR::Register *temp2Reg = addConstantToLong(node, src1Reg, -(value+1), NULL, cg);
-         temp1Reg = cg->allocateRegister();
-         killTemp1 = true;
-         if (value >= 0)
-            {
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::OR, node, temp1Reg, temp2Reg, src1Reg);
-            }
-         else // if (value < 0)
-            {
-            generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, temp1Reg, temp2Reg, src1Reg);
-            }
-         cg->stopUsingRegister(temp2Reg);
-         }
-      generateShiftRightLogicalImmediateLong(cg, node, trgReg, temp1Reg, 63);
-      if (killTemp1)
-         cg->stopUsingRegister(temp1Reg);
-      }
-   else
-      {
-      TR::Register   *src2Reg = cg->evaluate(secondChild);
-      TR::TreeEvaluator::genBranchSequence(node, src1Reg, src2Reg, trgReg, TR::InstOpCode::ble, TR::InstOpCode::cmp8, cg);
-      }
-   node->setRegister(trgReg);
-   cg->decReferenceCount(firstChild);
-   cg->decReferenceCount(secondChild);
-   return trgReg;
+TR::Register *OMR::Power::TreeEvaluator::fcmpequEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return floatCompareEvaluator(node, CompareInfo(CompareCondition::eq, TR::Float, true), cg);
    }
 
-TR::Register *OMR::Power::TreeEvaluator::lcmpgeEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+TR::Register *OMR::Power::TreeEvaluator::fcmpneEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   if (cg->comp()->target().is32Bit())
-     return compareLongAndSetOrderedBoolean(TR::InstOpCode::cmp4, TR::InstOpCode::bge, node, cg);
+   return floatCompareEvaluator(node, CompareInfo(CompareCondition::ne, TR::Float, false), cg);
```

Comment:
This kind of refactoring is really dangerous for Current Releases.  I think the isUnsignedOrUnordered flag should be true instead of false.  Unordered surely implies not-equal:  i.e., if two operands are unordered, their compare for not-equal should return true.  But, if the flag is false, it ends up isReversed being true and flag being false.  They are not equal, such that FU bit will be or-ed into EQ bit, so that they compare equal. That is completely wrong.

Reply:
I have no doubt there are a few such cases in this code.

Reply:
This code as written is correct. You seem to have misunderstood the semantics of the `fcmpne` opcode itself: this opcode is explicitly specified to return 0 for unordered operands [1] (all FP compare opcodes that return 1 for unordered operands have a `u` suffix on their name) and is the logical complement of `fcmpequ` not `fcmpeq`. The previous implementation in the Power codegen was incorrect and there is actually already a bug report (#5152) linked in one of the commit messages regarding that. The opcode that does implement the behaviour you're thinking about is `fcmpneu` and use of `fcmpne` is actually exceedingly rare.

[1] The properties for this opcode are defined here and its flags do not contain `ILProp3::CompareTrueIfUnordered`: https://github.com/eclipse/omr/blob/7970637018048a2bb00936855ae1c180861ac36b/compiler/il/OMRILOpCodeProperties.hpp#L3672-L3686